marklogic-data-hub/src/test/java/com/marklogic/hub/flow/impl/FlowMigratorTest.java

```diff
@@ -0,0 +1,63 @@
+package com.marklogic.hub.flow.impl;
+
+import com.marklogic.hub.AbstractHubCoreTest;
+import com.marklogic.hub.HubConfig;
+import com.marklogic.hub.HubProject;
+import org.apache.commons.io.FileUtils;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.skyscreamer.jsonassert.JSONAssert;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Arrays;
+
+class FlowMigratorTest extends AbstractHubCoreTest {
+    @BeforeEach
+    void setUp() {
+        try {
+            FileUtils.copyDirectory(getResourceFile("flow-migration-test/flows"), getHubConfig().getFlowsDir().toFile());
+            FileUtils.copyDirectory(getResourceFile("flow-migration-test/mappings"), getHubConfig().getHubMappingsDir().toFile());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @AfterEach
+    void tearDown() {
+        deleteProjectDir();
+    }
+
+    @Test
+    void migrateFlows() {
+        HubConfig hubConfig = getHubConfig();
+        HubProject hubProject = hubConfig.getHubProject();
+        FlowMigrator flowMigrator = new FlowMigrator(getHubConfig());
+        flowMigrator.migrateFlows();
+        Path migratedFlows = hubProject.getProjectDir().resolve("migrated-flows");
+
+        Assertions.assertTrue(migratedFlows.toFile().exists());
+        Assertions.assertTrue(migratedFlows.resolve("flows").toFile().listFiles().length > 0);
+        Assertions.assertTrue(migratedFlows.resolve("mappings").toFile().listFiles().length > 0);
+
+        verifyFiles(getResourceFile("flow-migration-test/keys/flows"), hubProject.getFlowsDir().toFile());
+        verifyFiles(getResourceFile("flow-migration-test/keys/steps/ingestion"), hubProject.getProjectDir().resolve("steps").resolve("ingestion").toFile());
+        verifyFiles(getResourceFile("flow-migration-test/keys/steps/mapping"), hubProject.getProjectDir().resolve("steps").resolve("mapping").toFile());
+    }
+
+    private void verifyFiles(File expectedDir, File actualDir) {
```

Comment:
I ran the test locally (it passed), but I think two changes should be made to the test:

1. Narrow down the test files to the bare minimum of steps needed to verify that everything is working
1. Add assertions to verify that the step references are correct in every flow, and that each step document was created correctly

For the first item above - the ingestion_only-flow.flow.json file is testing 7 identical scenarios, because every one of the 7 steps is the same from a QA perspective. So we only need one non-custom ingestion step that has a fileLocations. We also need a non-custom ingestion step that doesn't have inputFileType in it so we can verify that the migrator adds sourceFormat=json. 

As for the logic of - "Should this step be migrated?" - the most effective way to do that is to make "stepRequiresMigration" a protected method and that pass different kinds of steps into it, verifying that the result is true or false. 

For the second item - I think it's best to make an assertion on every stepId in the migrated flow document, and also on every property of every migrated step. We specifically want to verify that everything under "options" was migrated to top-level properties; that "fileLocations" was handled correctly; and that "options/mapping" was *not* migrated (it currently is being migrated, but should not be). 

I also noticed that a "namespaces: {}" was added, even though no namespaces exist in the mappings. If that's something we need to account for, then we need a mapping artifact that has namespaces populated so we can verify that the special handling for it worked. 

Note that we really do not want any issues during this migration process; the last thing we want for new users on 5.3 is to migrate their flows and then run into a problem that requires involving Support and makes them think that 5.3 doesn't work. So we need to be certain to cover every scenario and ensure every line of the migrated flows/steps is correct.